Automattic/jetpack
# CRM: Correcting secondary address field placeholders

Resolve: https://github.com/Automattic/zero-bs-crm/issues/1632#issuecomment-1461670297

## Proposed changes:

* This PR fixes an issue mentioned here - https://github.com/Automattic/zero-bs-crm/issues/1632#issuecomment-1461670297 - that secondary address placeholders weren't working.
* Noting here as well that the change was made in the` load_from_object_models` function in the[ jpcrm-templating-placeholders.php file](https://github.com/Automattic/jetpack/blob/trunk/projects/plugins/crm/includes/jpcrm-templating-placeholders.php#L1068) so that it would affect both the placeholder templates dropdown (for example in the mail template page), and also on the placeholder map. To change the output only on the placeholder map the change could be made within [placeholder-map.page.php](https://github.com/Automattic/jetpack/blob/trunk/projects/plugins/crm/admin/system/placeholder-map.page.php) directly.
* The issue itself stems to changes in secondary address field names between DAL versions. While `sec` is the [currently defined prefix for secondary addresses](https://github.com/Automattic/jetpack/blob/trunk/projects/plugins/crm/includes/ZeroBSCRM.DAL3.Fields.php#L781) , the issue with secondary address fields is something that is worked around in other places such as [here for exporting files](https://github.com/Automattic/jetpack/blob/trunk/projects/plugins/crm/includes/ZeroBSCRM.DAL3.Export.php#L341), and an exception made for [formatting addresses here](https://github.com/Automattic/jetpack/blob/trunk/projects/plugins/crm/includes/ZeroBSCRM.DAL3.php#L6507).
* I am still familiarizing myself with the DAL and how second address keys have changed for DAL3 so it's possible there may be somewhere else closer to the 'root' where these changes need to happen to avoid all exceptions, not just placeholders. I'm unsure about that at the moment though so would love some feedback.

### Other information:


## Jetpack product discussion

https://github.com/Automattic/zero-bs-crm/issues/1632#issuecomment-1461670297

## Does this pull request change what data or activity we track or use?

No.

## Testing instructions:

* To test the issue out, on trunk or with CRM installed up to 5.5.3, visit the placeholder map and copy the placeholder for a secondary address field, such as `##CONTACT-SECADDR1##`.
* Under Quote Templates, add a new template using that placeholder.
* Create a contact and make sure to fill out the first line of the secondary address.
* Create a quote for that new contact using the Quote Builder, making sure to select the template you've just created.
* The output should show the unrendered placeholder.
* To test this, using the Jetpack Beta plugin on a test site with this branch applied or testing locally with the branch checked out, visit the placeholder map page and copy the placeholder for the same secondary address field as before, which in the above example will now be ~`##CONTACT-SECADDR-ADDR1##`~ `##CONTACT-SECADDR_ADDR1##`.
* Replace the earlier quote template with the new placeholder, and for the same contact create a new quote.
* The output should show what was added in that first secondary address field.
* To test the placeholder dropdown for mail templates, visit CRM Settings / Mail Settings and choose Email Templates. From any template, click on 'Insert Placeholder', and scroll down to find 'Contact Address Line 1 (Second Address)'. Confirm the placeholder it suggests is the same as on the placeholder map (in this case ~`##CONTACT-SECADDR-ADDR1##`~ `##CONTACT-SECADDR_ADDR1##`.).
* Side note - the company second address placeholders don't need to change as they work as is (eg. testing adding invoices and downloading pdfs, after adding the placeholder to invoice-pdf.html file).
